src/serializer/ResidualHeapVisitor.js

```diff
@@ -487,30 +495,90 @@ export class ResidualHeapVisitor {
         this.referencedDeclaredValues.add(value);
       },
       recordDelayedEntry: (entry: GeneratorEntry) => {
-        this.delayedVisitGeneratorEntries.push({ generator, entry });
+        this.delayedVisitGeneratorEntries.push({ baseContext, generator, entry });
       },
     };
   }
 
-  visitGenerator(generator: Generator): void {
+  visitGenerator(generator: Generator, afterGeneratorVisit?: () => void): void {
     this._withScope(generator, () => {
-      generator.visit(this.createGeneratorVisitCallbacks(generator));
+      generator.visit(this.createGeneratorVisitCallbacks(generator, this.baseContext));
+      if (afterGeneratorVisit) afterGeneratorVisit();
     });
   }
 
-  visitRoots(): void {
-    this.visitGenerator(this.realmGenerator);
+  visitRoots(additionalFunctionValuesAndEffects: Map<FunctionValue, Effects>): void {
+    let generator = this.realm.generator;
+    invariant(generator);
+    this.visitGenerator(generator);
     for (let [, moduleValue] of this.modules.initializedModules) this.visitValue(moduleValue);
-    // Do a fixpoint over all pure generator entries to make sure that we visit
-    // arguments of only BodyEntries that are required by some other residual value
-    let oldDelayedEntries = [];
-    while (oldDelayedEntries.length !== this.delayedVisitGeneratorEntries.length) {
-      oldDelayedEntries = this.delayedVisitGeneratorEntries;
-      this.delayedVisitGeneratorEntries = [];
-      for (let { generator, entry } of oldDelayedEntries)
-        this._withScope(generator, () => {
-          generator.visitEntry(entry, this.createGeneratorVisitCallbacks(generator));
-        });
-    }
+    this.realm.evaluateAndRevertInGlobalEnv(() => {
+      if (additionalFunctionValuesAndEffects.size) {
+        for (let [functionValue, effects] of additionalFunctionValuesAndEffects.entries()) {
+          let [r, g, ob, pb: Map<PropertyBinding, void | Descriptor>, co] = effects;
+          // Need to do this fixup because otherwise we will skip over this function's
```

Comment:
Instead of this fixup here, can you set the right generator before the additional functions are interpreted, so that this new generator will have the right parent to begin with?